core/ide/che-core-ide-jseditor/src/main/java/org/eclipse/che/ide/jseditor/client/debug/BreakpointManagerImpl.java

```diff
@@ -186,52 +191,43 @@ private void deleteBreakpoints(final Set<String> paths) {
     /**
      * Adds breakpoint to the list and JVM.
      */
-    private void addBreakpoint(final VirtualFile activeFile,
-                               final int lineNumber) {
-
-        String path = activeFile.getPath();
-        Breakpoint breakpoint = new Breakpoint(Type.BREAKPOINT,
-                                               lineNumber,
-                                               activeFile.getPath(),
-                                               activeFile,
-                                               false);
-
-        List<Breakpoint> pathBreakpoints = breakpoints.get(path);
+    private void addBreakpoint(final Breakpoint breakpoint) {
+        List<Breakpoint> pathBreakpoints = breakpoints.get(breakpoint.getPath());
         if (pathBreakpoints == null) {
             pathBreakpoints = new ArrayList<>();
-            breakpoints.put(path, pathBreakpoints);
+            breakpoints.put(breakpoint.getPath(), pathBreakpoints);
         }
 
         if (!pathBreakpoints.contains(breakpoint)) {
             pathBreakpoints.add(breakpoint);
         }
+        preserveBreakpoints();
 
-        final BreakpointRenderer breakpointRenderer = getBreakpointRendererForFile(path);
+        final BreakpointRenderer breakpointRenderer = getBreakpointRendererForFile(breakpoint.getPath());
         if (breakpointRenderer != null) {
-            breakpointRenderer.addBreakpointMark(lineNumber, new LineChangeAction() {
+            breakpointRenderer.addBreakpointMark(breakpoint.getLineNumber(), new LineChangeAction() {
                 @Override
                 public void onLineChange(VirtualFile file, int firstLine, int linesAdded, int linesRemoved) {
                     BreakpointManagerImpl.this.onLineChange(file, firstLine, linesAdded, linesRemoved);
                 }
             });
-            breakpointRenderer.setBreakpointActive(lineNumber, breakpoint.isActive());
+            breakpointRenderer.setBreakpointActive(breakpoint.getLineNumber(), breakpoint.isActive());
         }
-        preserveBreakpoints();
 
         for (BreakpointManagerObserver observer : observers) {
             observer.onBreakpointAdded(breakpoint);
         }
 
         Debugger debugger = debuggerManager.getActiveDebugger();
         if (debugger != null) {
-            debugger.addBreakpoint(activeFile, lineNumber);
+            debugger.addBreakpoint(breakpoint.getFile(), breakpoint.getLineNumber());
         }
     }
 
     /**
      * Indicates if line of code to add breakpoint at is executable.
      */
-    private boolean isCodeExecutable(final VirtualFile activeFile, int lineNumber) {
+    private boolean isLineNotEmpty(final VirtualFile activeFile, int lineNumber) {
         EditorPartPresenter editor = getEditorForFile(activeFile.getPath());
         if (editor instanceof EmbeddedTextEditorPresenter) {
             Document document = ((EmbeddedTextEditorPresenter)editor).getDocument();
```

Comment:
EmbeddedTextEditorPresenter this is concreate implemenation. We should use interface org.eclipse.che.ide.jseditor.client.texteditor.TextEditor instead of this.